[no-extraneous-import] Error message reports the "@types/blah" package instead of the "blah" package

Here's a minimal repro: https://github.com/cakoose/no-extraneous-dependencies-reporting-types-package-bug

## Summary of repo contents

index.ts

```
import qs from 'qs';
...
```

package.json (I forget to add a dependency on "qs", but it gets included indirectly via "express")

```
    "dependencies": {
        "express": "^4.21.2"
    },
    "devDependencies": {
        "@types/qs": "^6.9.7",
        ...
    }
```

eslint.config.mjs:

```
    rules: {
        'import/no-extraneous-dependencies': ['error', {
            devDependencies: false,
            peerDependencies: false,
        }],
    },
```

## To reproduce

npm install
npm run lint

```
 .../index.ts
   1:1  error  '@types/qs' should be listed in the project's dependencies, not devDependencies
 import/no-extraneous-dependencies
```

Problem: It's reporting "@types/qs" instead of "qs".

This slight reporting issue really threw me off. I was trying all kinds of weird things to figure out why it was reporting the "@types/qs" package, only to realize that I didn't include "qs" in my dependencies list.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -15,6 +15,7 @@ This change log adheres to standards from [Keep a CHANGELOG](https://keepachange
 - ExportMap: resolve export * as ns re-exports under modern parsers ([#3250], thanks [@rasmi] [@JounQin] [@butterybread])
 - [`order`]: make the alphabetize comparator transitive for sibling/parent imports in the same group, so the autofix converges under Node 25's updated V8 sort ([#3235], thanks [@Sekhmet])
 - [`no-duplicates`]: remove extra space when merging import specifiers ([#3246], thanks [@mixelburg] [@shastaxc])
+- [`no-extraneous-dependencies`]: report runtime package name, not `@types/*`, when an import resolves via a `@types` devDependency ([#3208], thanks [@morgan-coded] [@cakoose] [@JounQin])
 
 ### Changed
 - [Docs] `no-unused-modules`: Fix docs of `ignoreUnusedTypeExports` option ([#3233], thanks [@ej612])
@@ -1201,6 +1202,7 @@ for info on changes for earlier releases.
 [#3236]: https://github.com/import-js/eslint-plugin-import/pull/3236
 [#3235]: https://github.com/import-js/eslint-plugin-import/issues/3235
 [#3233]: https://github.com/import-js/eslint-plugin-import/pull/3233
+[#3208]: https://github.com/import-js/eslint-plugin-import/issues/3208
 [#3195]: https://github.com/import-js/eslint-plugin-import/issues/3195
 [#3191]: https://github.com/import-js/eslint-plugin-import/pull/3191
 [#3173]: https://github.com/import-js/eslint-plugin-import/pull/3173
@@ -1863,6 +1865,7 @@ for info on changes for earlier releases.
 [@brendo]: https://github.com/brendo
 [@brettz9]: https://github.com/brettz9
 [@butterybread]: https://github.com/butterybread
+[@cakoose]: https://github.com/cakoose
 [@chabb]: https://github.com/chabb
 [@Chamion]: https://github.com/Chamion
 [@charlessuh]: https://github.com/charlessuh
@@ -2018,6 +2021,7 @@ for info on changes for earlier releases.
 [@mixelburg]: https://github.com/mixelburg
 [@mlenser]: https://github.com/mlenser
 [@moeriki]: https://github.com/moeriki
+[@morgan-coded]: https://github.com/morgan-coded
 [@mpalmer685]: https://github.com/mpalmer685
 [@mpint]: https://github.com/mpint
 [@mplewis]: https://github.com/mplewis

--- a/src/rules/no-extraneous-dependencies.js
+++ b/src/rules/no-extraneous-dependencies.js
@@ -147,6 +147,13 @@ function getModuleRealName(resolved) {
   return getFilePackageName(resolved);
 }
 
+// derive the runtime package a DefinitelyTyped `@types/*` package provides types for,
+// mirroring TypeScript's `getPackageNameFromTypesPackageName`: `@types/babel__core` -> `@babel/core`
+function runtimePackageNameFromTypesPackageName(packageName) {
+  const innerPackageName = packageName.slice('@types/'.length);
+  return innerPackageName.includes('__') ? `@${innerPackageName.replace('__', '/')}` : innerPackageName;
+}
+
 function checkDependencyDeclaration(deps, packageName, declarationStatus) {
   const newDeclarationStatus = declarationStatus || {
     isInDeps: false,
@@ -231,6 +238,11 @@ function reportIfMissing(context, deps, depsOptions, node, name, moduleSystem) {
     ) {
       return;
     }
+  }
+
+  if (realPackageName && realPackageName.startsWith('@types/')) {
+    context.report(node, missingErrorMessage(runtimePackageNameFromTypesPackageName(realPackageName)));
+    return;
   }
 
   if (declarationStatus.isInDevDeps && !depsOptions.allowDevDeps) {

--- a/tests/files/with-resolved-types-dependencies/foo.js
+++ b/tests/files/with-resolved-types-dependencies/foo.js
@@ -1,0 +1,3 @@
+import qs from 'qs';
+
+export default qs;

--- a/tests/files/with-resolved-types-dependencies/package.json
+++ b/tests/files/with-resolved-types-dependencies/package.json
@@ -1,0 +1,6 @@
+{
+  "devDependencies": {
+    "@types/babel__core": "*",
+    "@types/qs": "*"
+  }
+}

--- a/tests/src/rules/no-extraneous-dependencies.js
+++ b/tests/src/rules/no-extraneous-dependencies.js
@@ -20,6 +20,7 @@ const packageFileWithSyntaxErrorMessage = (() => {
 })();
 const packageDirWithFlowTyped = path.join(__dirname, '../../files/with-flow-typed');
 const packageDirWithTypescriptDevDependencies = path.join(__dirname, '../../files/with-typescript-dev-dependencies');
+const packageDirWithResolvedTypesDependencies = path.join(__dirname, '../../files/with-resolved-types-dependencies');
 const packageDirMonoRepoRoot = path.join(__dirname, '../../files/monorepo');
 const packageDirMonoRepoWithNested = path.join(__dirname, '../../files/monorepo/packages/nested-package');
 const packageDirWithEmpty = path.join(__dirname, '../../files/empty');
@@ -402,6 +403,27 @@ ruleTester.run('no-extraneous-dependencies', rule, {
       errors: [{
         // missing dependency is chai not alias
         message: "'chai' should be listed in the project's dependencies. Run 'npm i -S chai' to add it",
+      }],
+    }),
+
+    // https://github.com/import-js/eslint-plugin-import/issues/3208
+    // when `qs` resolves only through the `@types/qs` devDependency, the missing
+    // runtime package is `qs`, not the `@types/qs` package the resolver landed on
+    test({
+      code: 'import qs from "qs";',
+      filename: path.join(packageDirWithResolvedTypesDependencies, 'foo.js'),
+      options: [{ packageDir: packageDirWithResolvedTypesDependencies, devDependencies: false }],
+      errors: [{
+        message: "'qs' should be listed in the project's dependencies. Run 'npm i -S qs' to add it",
+      }],
+    }),
+
+    test({
+      code: 'import babel from "@babel/core";',
+      filename: path.join(packageDirWithResolvedTypesDependencies, 'foo.js'),
+      options: [{ packageDir: packageDirWithResolvedTypesDependencies, devDependencies: false }],
+      errors: [{
+        message: "'@babel/core' should be listed in the project's dependencies. Run 'npm i -S @babel/core' to add it",
       }],
     }),
 

